File: src/core/sdk/python/acgs2_sdk/services/api_gateway.py
"""
ACGS-2 API Gateway Service
Constitutional Hash: cdd01ef066bc6cf2
"""

from typing import TYPE_CHECKING, Any, Dict, List, Union, Optional

try:
    from src.core.shared.types import JSONDict, JSONValue
except ImportError:
    JSONValue = Union[str, int, float, bool, None, Dict[str, Any], List[Any]]  # type: ignore[misc]
    JSONDict = Dict[str, JSONValue]  # type: ignore[misc]

if TYPE_CHECKING:
    from acgs2_sdk.client import ACGS2Client


class APIGatewayService:
    """Service for interacting with the API Gateway endpoints."""

    def __init__(self, client: "ACGS2Client") -> None:
        """Initialize the API gateway service.

        Args:
            client: ACGS2 client instance
        """
        self._client = client

    async def health_check(self) -> JSONDict:
        """Check API gateway health.

        Returns:
            Health check response
        """
        return await self._client.get("/health")

    async def submit_feedback(
        self,
        user_id: str,
        category: str,
        rating: int,
        title: str,
        description: str = "",
        metadata: Optional[JSONDict] = None,
    ) -> JSONDict:
        """Submit user feedback.

        Args:
            user_id: User identifier
            category: Feedback category (bug, feature, general)
            rating: Rating from 1-5
            title: Feedback title
            description: Feedback description
            metadata: Additional metadata

        Returns:
            Feedback submission response
        """
        data = {
            "user_id": user_id,
            "category": category,
            "rating": rating,
            "title": title,
            "description": description,
        }

        if metadata:
            data["metadata"] = metadata

        return await self._client.post("/feedback", json=data)

    async def get_feedback_stats(self) -> JSONDict:
        """Get feedback statistics.

        Returns:
            Feedback statistics
        """
        return await self._client.get("/feedback/stats")

    async def list_services(self) -> JSONDict:
        """List available services.

        Returns:
            Services information
        """
        return await self._client.get("/services")

File: src/core/sdk/go/api_gateway.go
package sdk

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
)

// APIGatewayService provides methods for interacting with the API Gateway
type APIGatewayService struct {
	client *ACGS2Client
}

// NewAPIGatewayService creates a new APIGatewayService instance
func NewAPIGatewayService(client *ACGS2Client) *APIGatewayService {
	return &APIGatewayService{
		client: client,
	}
}

// HealthCheck performs a health check on the API gateway
func (s *APIGatewayService) HealthCheck(ctx context.Context) (*HealthCheckResponse, error) {
	url := fmt.Sprintf("%s/health", s.client.config.BaseURL)

	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	resp, err := s.client.doRequest(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var health HealthCheckResponse
	if err := json.NewDecoder(resp.Body).Decode(&health); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	return &health, nil
}

// SubmitFeedback submits user feedback
func (s *APIGatewayService) SubmitFeedback(ctx context.Context, req FeedbackRequest) (*FeedbackResponse, error) {
	url := fmt.Sprintf("%s/feedback", s.client.config.BaseURL)

	requestData := map[string]interface{}{
		"userId":             req.UserID,
		"category":           req.Category,
		"rating":             req.Rating,
		"title":              req.Title,
		"constitutionalHash": ConstitutionalHash,
	}

	if req.Description != nil {
		requestData["description"] = *req.Description
	}
	if req.Metadata != nil {
		requestData["metadata"] = req.Metadata
	}

	jsonData, err := json.Marshal(requestData)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	httpReq, err := http.NewRequestWithContext(ctx, "POST", url, bytes.NewBuffer(jsonData))
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}
	httpReq.Header.Set("Content-Type", "application/json")

	resp, err := s.client.doRequest(httpReq)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var feedbackResp FeedbackResponse
	if err := json.NewDecoder(resp.Body).Decode(&feedbackResp); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	return &feedbackResp, nil
}

// GetFeedbackStats retrieves feedback statistics
func (s *APIGatewayService) GetFeedbackStats(ctx context.Context) (*FeedbackStats, error) {
	url := fmt.Sprintf("%s/feedback/stats", s.client.config.BaseURL)

	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	resp, err := s.client.doRequest(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var stats FeedbackStats
	if err := json.NewDecoder(resp.Body).Decode(&stats); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	return &stats, nil
}

// ListServices retrieves available services information
func (s *APIGatewayService) ListServices(ctx context.Context) (*ServicesResponse, error) {
	url := fmt.Sprintf("%s/services", s.client.config.BaseURL)

	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}

	resp, err := s.client.doRequest(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var servicesResp ServicesResponse
	if err := json.NewDecoder(resp.Body).Decode(&servicesResp); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	return &servicesResp, nil
}
